File: js/euler53.js
function euler53(){
    var res = 0;
    for (let n = 10; n <= 100; n++){
        let ncr = n;
        // nC1 = n
        // nC2 = (n^2 - n) / 2 = n * (n - 1) / 2
        // nC3 = (n^3 - 3n^2 + 2n) = n * (n - 1) / 2 * (n - 2) / 3
		// nCr = nC(r - 1) * (n - r + 1) / r
        for (let r = 2; r <= n / 2; r++){
            ncr *= (n - r + 1) / r;
            if (ncr > 1_000_000){
                // n choose r = n choose (n - r)
                // [r, n - r] = [r + 0, r + n - 2r]
                // [r, n - r] has n - 2r + 1 values
                res += n - 2 * r + 1;
                break;
            }
        }
    }
    return res;
}

console.log(euler53());

File: go/euler53.go
package main
import "fmt"

func main() {
    res := 0
    // nCr first exceeds 1 million when n = 23
    for n := 23; n <= 100; n++ {
        ncr := 1
        // nC1 = n
        // nC2 = n * (n - 1) / 2
        // nC3 = n * (n - 1) * (n - 2) / (2 * 3)
        // nCr = nC(r - 1) * (n - r + 1) / r
        for r := 1; r <= n; r++ {
            ncr = ncr * (n - r + 1) / r
            // first r in row n that makes nCr exceed 1 million
            // combinatorics are symmetric, nCr = nC(n - r)
            if ncr > 1000000 {
                // size of range [r, n - r]: n - r - r + 1 = n - 2r + 1
                res += n - 2 * r + 1
                break
            }
        }
    }
    fmt.Println(res)
}
